pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/token_counter.go

```diff
@@ -0,0 +1,533 @@
+	defer func() {
+		// Signal the next waiter if we have tokens available upon returning.
+		if haveTokens {
+			wh.b.mu.counters[wh.wc].signal()
+		}
+	}()
+
+	if tokens > 0 {
+		granted = wh.b.TryDeduct(context.Background(), wh.wc, tokens)
+	}
+
+	return granted, wh.b.tokens(wh.wc) > 0
+}
+
+// WaitForEval waits for a quorum of handles to be signaled and have tokens
+// available, including all the required wait handles. The caller can provide a
+// refresh channel, which when signaled will cause the function to return
+// RefreshWaitSignaled, allowing the caller to retry waiting with updated
+// handles.
+//
+// If the required quorum and required wait handles are signaled and have
+// tokens available, the function checks the availability of tokens for all
+// handles. If not all required handles have available tokens or the available
+// count is less than the required quorum, RefreshWaitSignaled is also
+// returned.
+func WaitForEval(
+	ctx context.Context,
+	refreshWaitCh <-chan struct{},
+	handles []tokenWaitingHandleInfo,
+	requiredQuorum int,
+	scratch []reflect.SelectCase,
+) (state WaitEndState, scratch2 []reflect.SelectCase) {
+	scratch = scratch[:0]
+	if len(handles) < requiredQuorum || requiredQuorum == 0 {
+		panic(errors.AssertionFailedf("invalid arguments to WaitForEval"))
+	}
+
+	scratch = append(scratch,
+		reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(ctx.Done())})
+	scratch = append(scratch,
+		reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(refreshWaitCh)})
+
+	requiredWaitCount := 0
+	for _, h := range handles {
+		if h.requiredWait {
+			requiredWaitCount++
+		}
+		scratch = append(scratch,
+			reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(h.handle.WaitChannel())})
+	}
+	// We track the original length of the scratch slice and also the current
+	// length, where after each successful iteration we decrement the current
+	// length.
+	m := len(scratch)
+	signaledCount := 0
+
+	// Wait for at least a quorumCount of handles to be signaled which also have
+	// tokens and additionally, all of the required wait handles to have signaled
+	// and have tokens available.
+	for signaledCount < requiredQuorum || requiredWaitCount > 0 {
+		chosen, _, _ := reflect.Select(scratch)
+		switch chosen {
+		case 0:
+			return ContextCanceled, scratch
+		case 1:
+			return RefreshWaitSignaled, scratch
+		default:
+			handleInfo := handles[chosen-2]
+			signaledCount++
+			if handleInfo.requiredWait {
+				requiredWaitCount--
+			}
+			m--
+			scratch[chosen], scratch[m] = scratch[m], scratch[chosen]
+			scratch = scratch[:m]
+			handles[chosen-2], handles[m-2] = handles[m-2], handles[chosen-2]
+		}
+	}
+
+	availableCount := 0
+	allRequiredAvailable := true
+	// Check whether the handle has available tokens, if not we keep the
+	// select case and keep on waiting.
+	// TODO(kvoli): Currently we are only trying to deduct and signal the next
```

Comment:
Chatting w/ @sumeerbhola:

We should unblock the next waiter immediately. Utilize the normal back-pressure the goroutine scheduling is giving us. 

Refresh is only being closed when the replica set changes. Otherwise, we expect this to successfully wait or get cancelled.

Worst case burst of GOMAXPROCS acquiring tokens when signaled. 

TODO(kvoli): update code to try deducting immediately after being signaled on a handle's channel